go/store/nbs/archive_writer.go

```diff
@@ -0,0 +1,397 @@
+    does not write anything to disk yet.
+ 3. Repeat steps 1 and 2 as necessary. You can interleave them, but all chunks must be staged before the next step.
+ 4. finalizeByteSpans: At this point, all byte spans have been written out, and the checksum for the data block
+    is calculated. No more byte spans can be written after this step.
+ 5. writeIndex: Write the index to the archive. This will do all the work of writing the byte span map, prefix map,
+    chunk references, and suffixes. Index checksum is calculated at the end of this step.
+ 6. writeMetadata: Write the metadataSpan to the archive. Calculate the metadataSpan checksum at the end of this step.
+ 7. writeFooter: Write the footer to the archive. This will write out the index length, byte span count, chunk count.
+ 8. flushToFile: Write the archive to disk and move into its new home.
+
+When all of these steps have been completed without error, the ByteSink used to create the writer can be flushed and closed
+to complete the archive writing process.
+*/
+
+func newArchiveWriterWithSink(bs ByteSink) *archiveWriter {
+	hbs := NewSHA512HashingByteSink(bs)
+	return &archiveWriter{output: hbs, seenChunks: hash.HashSet{}}
+}
+
+// writeByteSpan writes a byte span to the archive, returning the ByteSpan ID if the write was successful. Note
+// that writing an empty byte span is a no-op and will return 0. Also, the slice passed in is copied, so the caller
+// can reuse the slice after this call.
+func (aw *archiveWriter) writeByteSpan(b []byte) (uint32, error) {
+	if aw.workflowStage != stageByteSpan {
+		return 0, fmt.Errorf("Runtime error: writeByteSpan called out of order")
+	}
+
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	offset := aw.bytesWritten
+
+	written, err := aw.output.Write(b)
+	if err != nil {
+		return 0, err
+	}
+	if written != len(b) {
+		return 0, io.ErrShortWrite
+	}
+	aw.bytesWritten += uint64(written)
+
+	aw.stagedBytes = append(aw.stagedBytes, byteSpan{offset, uint64(written)})
+
+	return uint32(len(aw.stagedBytes)), nil
+}
+
+func (aw *archiveWriter) chunkSeen(h hash.Hash) bool {
+	return aw.seenChunks.Has(h)
+}
+
+func (aw *archiveWriter) stageChunk(hash hash.Hash, dictionary, data uint32) error {
+	if aw.workflowStage != stageByteSpan {
+		return fmt.Errorf("Runtime error: stageChunk called out of order")
+	}
+
+	if data == 0 || data > uint32(len(aw.stagedBytes)) {
+		return ErrInvalidChunkRange
+	}
+	if aw.seenChunks.Has(hash) {
+		return ErrDuplicateChunkWritten
+	}
+	aw.seenChunks.Insert(hash)
+
+	if dictionary > uint32(len(aw.stagedBytes)) {
+		return ErrInvalidDictionaryRange
+	}
+
+	aw.stagedChunks = append(aw.stagedChunks, stagedChunkRef{hash, dictionary, data})
+	return nil
+}
+
+func (scrs stagedChunkRefSlice) Len() int {
+	return len(scrs)
+}
+func (scrs stagedChunkRefSlice) Less(i, j int) bool {
+	return scrs[i].hash.Prefix() < scrs[j].hash.Prefix()
```

Comment:
I think if just `bytes.Compare(scrs[i].hash[:], scrs[j].hash[:]) < 0` we get sorted fully on hash, which might be nicer in some ways, despite the fact that prefix and suffix are going to two different physical places in the file.